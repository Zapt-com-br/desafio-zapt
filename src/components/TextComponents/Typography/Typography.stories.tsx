import React from 'react';

import { Story, Meta } from '@storybook/react/types-6-0';
import Typography, { TypographyProps } from '.';

export default {
  title: 'Text Components/Typography',
  component: Typography,
  argTypes: {
    children: {
      type: 'string',
    },
  },
} as Meta;

export const TypographyDefault: Story<TypographyProps> = args => (
  <Typography {...args} />
);
