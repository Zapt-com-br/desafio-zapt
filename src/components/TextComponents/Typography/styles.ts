import styled, { css } from 'styled-components';

import { TypographyProps } from '.';

type StyledTextProps = Pick<TypographyProps, 'size' | 'color' | 'bold'>;

export const Text = styled.span<StyledTextProps>`
  ${({ theme, size, color, bold }) => css`
    font-weight: ${theme.font.weight[bold!]};
    color: ${theme.colors[color!]};
    font-size: ${theme.font.sizes[size!]};
  `}
`;
