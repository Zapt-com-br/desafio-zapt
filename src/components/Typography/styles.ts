import styled, { css, DefaultTheme } from 'styled-components'
import { TypographyProps } from '.'

const sizeModifiers = {
  xsmall: (theme: DefaultTheme) => css`
    font-size: ${theme.font.sizes.xsmall};
  `,
  small: (theme: DefaultTheme) => css`
    font-size: ${theme.font.sizes.small};
  `,
  medium: (theme: DefaultTheme) => css`
    font-size: ${theme.font.sizes.medium};
  `,
  large: (theme: DefaultTheme) => css`
    font-size: ${theme.font.sizes.large};
  `,
  xlarge: (theme: DefaultTheme) => css`
    font-size: ${theme.font.sizes.xlarge};
  `,
  huge: (theme: DefaultTheme) => css`
    font-size: ${theme.font.sizes.huge};
  `
}

const colorModifier = {
  primary: (theme: DefaultTheme) => css`
    color: ${theme.colors.primary};
  `,
  secondary: (theme: DefaultTheme) => css`
    color: ${theme.colors.secondary};
  `,
  lightGray: (theme: DefaultTheme) => css`
    color: ${theme.colors.lightGray};
  `,
  gray: (theme: DefaultTheme) => css`
    color: ${theme.colors.gray};
  `,
  lightBlue: (theme: DefaultTheme) => css`
    color: ${theme.colors.lightBlue};
  `,
  neutralBlue: (theme: DefaultTheme) => css`
    color: ${theme.colors.neutralBlue};
  `,
  darkBlue: (theme: DefaultTheme) => css`
    color: ${theme.colors.darkBlue};
  `,
  black: (theme: DefaultTheme) => css`
    color: ${theme.colors.black};
  `,
  white: (theme: DefaultTheme) => css`
    color: ${theme.colors.white};
  `
}

const weightModifier = {
  light: (theme: DefaultTheme) => css`
    font-weight: ${theme.font.weight.light};
  `,
  regular: (theme: DefaultTheme) => css`
    font-weight: ${theme.font.weight.regular};
  `,
  medium: (theme: DefaultTheme) => css`
    font-weight: ${theme.font.weight.medium};
  `
}

export const Typography = styled.h1<TypographyProps>`
  ${({ theme, color, size, weight }) => css`
    ${!!size && sizeModifiers[size](theme)}
    ${!!color && colorModifier[color](theme)}
    ${!!weight && weightModifier[weight](theme)}
  `}
`
