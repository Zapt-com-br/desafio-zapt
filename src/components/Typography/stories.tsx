import { Story, Meta } from '@storybook/react'
import Typography, { TypographyProps } from '.'

export default {
  title: 'Typography',
  component: Typography,
  argTypes: {
    children: {
      type: 'string'
    },
    as: {
      control: {
        type: 'select',
        options: ['h1', 'h2', 'p']
      }
    },
    weight: {
      control: {
        type: 'select',
        options: ['light', 'regular', 'medium']
      }
    },
    size: {
      control: {
        type: 'select',
        options: ['xsmall', 'small', 'medium', 'large', 'xlarge', 'huge']
      }
    },
    color: {
      control: {
        type: 'select',
        options: [
          'primary',
          'secondary',
          'white',
          'gray',
          'lightBlue',
          'neutralBlue',
          'darkBlue',
          'black'
        ]
      }
    }
  }
} as Meta

export const Default: Story<TypographyProps> = (args) => (
  <Typography {...args} />
)

Default.args = {
  children: 'Test'
}
