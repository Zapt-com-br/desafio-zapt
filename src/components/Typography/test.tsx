import { render, screen } from 'utils/test-utils'

import Typography from '.'

describe('<Typography />', () => {
  it('should render the Typography component with its children', () => {
    render(<Typography>Test</Typography>)

    expect(screen.getByText('Test')).toBeInTheDocument()
  })

  it('should render the default styles', () => {
    render(<Typography>Test</Typography>)

    expect(screen.getByRole('heading', { name: /test/i })).toHaveStyle({
      'font-size': '1.6rem',
      'font-weight': '400',
      color: '#000000'
    })
  })

  it('should render the xsmall size', () => {
    render(<Typography size="xsmall">Test 2</Typography>)

    expect(screen.getByRole('heading', { name: /test 2/i })).toHaveStyle({
      'font-size': '1.2rem'
    })
  })

  it(`should render the small size`, () => {
    render(<Typography size="small">Test 3</Typography>)

    expect(screen.getByRole('heading', { name: /test 3/i })).toHaveStyle({
      'font-size': '1.4rem'
    })
  })

  it(`should render the medium size`, () => {
    render(<Typography size="medium">Test 4</Typography>)

    expect(screen.getByRole('heading', { name: /test 4/i })).toHaveStyle({
      'font-size': '1.6rem'
    })
  })

  it(`should render the large size`, () => {
    render(<Typography size="large">Test 5</Typography>)

    expect(screen.getByRole('heading', { name: /test 5/i })).toHaveStyle({
      'font-size': '2rem'
    })
  })

  it(`should render the xlarge size`, () => {
    render(<Typography size="xlarge">Test 6</Typography>)

    expect(screen.getByRole('heading', { name: /test 6/i })).toHaveStyle({
      'font-size': '2.8rem'
    })
  })

  it(`should render the huge size`, () => {
    render(<Typography size="huge">Test 7</Typography>)

    expect(screen.getByRole('heading', { name: /test 7/i })).toHaveStyle({
      'font-size': '4rem'
    })
  })

  it(`should render the primary color`, () => {
    render(<Typography color="primary">Hello</Typography>)

    expect(screen.getByRole('heading', { name: /hello/i })).toHaveStyle({
      color: '#334BC8'
    })
  })

  it(`should render the secondary color`, () => {
    render(<Typography color="secondary">Hello</Typography>)

    expect(screen.getByRole('heading', { name: /hello/i })).toHaveStyle({
      color: '#FF5B58'
    })
  })

  it(`should render the lightGray color`, () => {
    render(<Typography color="lightGray">Hello</Typography>)

    expect(screen.getByRole('heading', { name: /hello/i })).toHaveStyle({
      color: '#FAFAFA'
    })
  })

  it(`should render the white color`, () => {
    render(<Typography color="white">Hello</Typography>)

    expect(screen.getByRole('heading', { name: /hello/i })).toHaveStyle({
      color: '#FFFFFF'
    })
  })

  it(`should render the gray color`, () => {
    render(<Typography color="gray">Hello</Typography>)

    expect(screen.getByRole('heading', { name: /hello/i })).toHaveStyle({
      color: '#7B7B7B'
    })
  })

  it(`should render the lightBlue color`, () => {
    render(<Typography color="lightBlue">Hello</Typography>)

    expect(screen.getByRole('heading', { name: /hello/i })).toHaveStyle({
      color: '#2196F3'
    })
  })

  it(`should render the neutralBlue color`, () => {
    render(<Typography color="neutralBlue">Hello</Typography>)

    expect(screen.getByRole('heading', { name: /hello/i })).toHaveStyle({
      color: '#011FBB'
    })
  })

  it(`should render the darkBlue color`, () => {
    render(<Typography color="darkBlue">Hello</Typography>)

    expect(screen.getByRole('heading', { name: /hello/i })).toHaveStyle({
      color: '#001C33'
    })
  })

  it(`should render the black color`, () => {
    render(<Typography color="black">Hello</Typography>)

    expect(screen.getByRole('heading', { name: /hello/i })).toHaveStyle({
      color: '#000000'
    })
  })

  it(`should render the light weight`, () => {
    render(<Typography weight="light">Hello</Typography>)

    expect(screen.getByRole('heading', { name: /hello/i })).toHaveStyle({
      'font-weight': '300'
    })
  })

  it(`should render the regular weight`, () => {
    render(<Typography weight="regular">Hello</Typography>)

    expect(screen.getByRole('heading', { name: /hello/i })).toHaveStyle({
      'font-weight': '400'
    })
  })

  it(`should render the medium weight`, () => {
    render(<Typography weight="medium">Hello</Typography>)

    expect(screen.getByRole('heading', { name: /hello/i })).toHaveStyle({
      'font-weight': '500'
    })
  })
})
