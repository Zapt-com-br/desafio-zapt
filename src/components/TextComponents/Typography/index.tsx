import React from 'react';
import * as Styled from './styles';

export type TypographyProps = {
  children: React.ReactNode;
  size?: 'xsmall' | 'small' | 'medium' | 'xmedium' | 'large';
  color?: 'black' | 'darkgray' | 'white' | 'primary';
  bold?: '400' | '500';
};

const Typography = ({
  children,
  size = 'medium',
  color = 'black',
  bold = '400',
}: TypographyProps) => (
  <Styled.Text bold={bold} color={color} size={size}>
    {children}
  </Styled.Text>
);

export default Typography;
