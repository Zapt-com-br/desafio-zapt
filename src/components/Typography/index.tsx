import * as S from './styles'

type TFontWeight = 'light' | 'regular' | 'medium'

type TFontSize = 'xsmall' | 'small' | 'medium' | 'large' | 'xlarge' | 'huge'

type TFontColor =
  | 'primary'
  | 'secondary'
  | 'lightGray'
  | 'gray'
  | 'lightBlue'
  | 'neutralBlue'
  | 'darkBlue'
  | 'black'
  | 'white'

export type TypographyProps = {
  children: React.ReactNode
  as?: React.ElementType
  weight?: TFontWeight
  size?: TFontSize
  color?: TFontColor
}

const Typography = ({
  children,
  as = 'h1',
  color = 'black',
  size = 'medium',
  weight = 'regular'
}: TypographyProps) => (
  <S.Typography as={as} color={color} size={size} weight={weight}>
    {children}
  </S.Typography>
)

export default Typography
