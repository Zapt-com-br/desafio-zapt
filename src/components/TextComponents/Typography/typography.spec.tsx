import React from 'react';
import { screen } from '@testing-library/react';
import { renderWithTheme } from 'helpers/renderWithThame';

import Typography from '.';

describe('<Typography />', () => {
  it('should render Typography correctly', () => {
    renderWithTheme(<Typography>text</Typography>);

    const element = screen.getByText(/text/i);

    expect(element).toBeInTheDocument();
    expect(element).toMatchSnapshot();
  });

  it('should render Typography whith color black for default', () => {
    renderWithTheme(<Typography>text</Typography>);

    const element = screen.getByText(/text/i);

    expect(element).toHaveStyleRule('color', '#000000');
    expect(element).toMatchSnapshot();
  });

  it('should render Typography whith colors correctly [darkgray]', () => {
    renderWithTheme(<Typography color="darkgray">text</Typography>);

    const element = screen.getByText(/text/i);

    expect(element).toHaveStyleRule('color', '#7B7B7B');
    expect(element).toMatchSnapshot();
  });

  it('should render Typography whith colors correctly [white]', () => {
    renderWithTheme(<Typography color="white">text</Typography>);

    const element = screen.getByText(/text/i);

    expect(element).toHaveStyleRule('color', '#ffffff');
    expect(element).toMatchSnapshot();
  });

  it('should render Typography whith colors correctly [primary]', () => {
    renderWithTheme(<Typography color="primary">text</Typography>);

    const element = screen.getByText(/text/i);

    expect(element).toHaveStyleRule('color', '#011FBB');
    expect(element).toMatchSnapshot();
  });

  it('should render Typography whith bold 400 for default', () => {
    renderWithTheme(<Typography>text</Typography>);

    const element = screen.getByText(/text/i);

    expect(element).toHaveStyleRule('font-weight', '400');
    expect(element).toMatchSnapshot();
  });

  it('should render Typography whith bold 500 if passed', () => {
    renderWithTheme(<Typography bold="500">text</Typography>);

    const element = screen.getByText(/text/i);

    expect(element).toHaveStyleRule('font-weight', '500');
    expect(element).toMatchSnapshot();
  });

  it('should render Typography whith font size medium for default', () => {
    renderWithTheme(<Typography>text</Typography>);

    const element = screen.getByText(/text/i);

    expect(element).toHaveStyleRule('font-size', '1.6rem');
    expect(element).toMatchSnapshot();
  });

  it('should render Typography whith font size xsmall is passed', () => {
    renderWithTheme(<Typography size="xsmall">text</Typography>);

    const element = screen.getByText(/text/i);

    expect(element).toHaveStyleRule('font-size', '1.2rem');
    expect(element).toMatchSnapshot();
  });

  it('should render Typography whith font size small is passed', () => {
    renderWithTheme(<Typography size="small">text</Typography>);

    const element = screen.getByText(/text/i);

    expect(element).toHaveStyleRule('font-size', '1.4rem');
    expect(element).toMatchSnapshot();
  });

  it('should render Typography whith font size xmedium is passed', () => {
    renderWithTheme(<Typography size="xmedium">text</Typography>);

    const element = screen.getByText(/text/i);

    expect(element).toHaveStyleRule('font-size', '2rem');
    expect(element).toMatchSnapshot();
  });

  it('should render Typography whith font size large is passed', () => {
    renderWithTheme(<Typography size="large">text</Typography>);

    const element = screen.getByText(/text/i);

    expect(element).toHaveStyleRule('font-size', '2.8rem');
    expect(element).toMatchSnapshot();
  });
});
