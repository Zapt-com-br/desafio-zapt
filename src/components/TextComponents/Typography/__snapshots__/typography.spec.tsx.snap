// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Typography /> should render Typography correctly 1`] = `
.c0 {
  font-weight: 400;
  color: #000000;
  font-size: 1.6rem;
}

<span
  class="c0"
  color="black"
>
  text
</span>
`;

exports[`<Typography /> should render Typography whith bold 400 for default 1`] = `
.c0 {
  font-weight: 400;
  color: #000000;
  font-size: 1.6rem;
}

<span
  class="c0"
  color="black"
>
  text
</span>
`;

exports[`<Typography /> should render Typography whith bold 500 if passed 1`] = `
.c0 {
  font-weight: 500;
  color: #000000;
  font-size: 1.6rem;
}

<span
  class="c0"
  color="black"
>
  text
</span>
`;

exports[`<Typography /> should render Typography whith color black for default 1`] = `
.c0 {
  font-weight: 400;
  color: #000000;
  font-size: 1.6rem;
}

<span
  class="c0"
  color="black"
>
  text
</span>
`;

exports[`<Typography /> should render Typography whith colors correctly [darkgray] 1`] = `
.c0 {
  font-weight: 400;
  color: #7B7B7B;
  font-size: 1.6rem;
}

<span
  class="c0"
  color="darkgray"
>
  text
</span>
`;

exports[`<Typography /> should render Typography whith colors correctly [primary] 1`] = `
.c0 {
  font-weight: 400;
  color: #011FBB;
  font-size: 1.6rem;
}

<span
  class="c0"
  color="primary"
>
  text
</span>
`;

exports[`<Typography /> should render Typography whith colors correctly [white] 1`] = `
.c0 {
  font-weight: 400;
  color: #ffffff;
  font-size: 1.6rem;
}

<span
  class="c0"
  color="white"
>
  text
</span>
`;

exports[`<Typography /> should render Typography whith font size large is passed 1`] = `
.c0 {
  font-weight: 400;
  color: #000000;
  font-size: 2.8rem;
}

<span
  class="c0"
  color="black"
>
  text
</span>
`;

exports[`<Typography /> should render Typography whith font size medium for default 1`] = `
.c0 {
  font-weight: 400;
  color: #000000;
  font-size: 1.6rem;
}

<span
  class="c0"
  color="black"
>
  text
</span>
`;

exports[`<Typography /> should render Typography whith font size small is passed 1`] = `
.c0 {
  font-weight: 400;
  color: #000000;
  font-size: 1.4rem;
}

<span
  class="c0"
  color="black"
>
  text
</span>
`;

exports[`<Typography /> should render Typography whith font size xmedium is passed 1`] = `
.c0 {
  font-weight: 400;
  color: #000000;
  font-size: 2rem;
}

<span
  class="c0"
  color="black"
>
  text
</span>
`;

exports[`<Typography /> should render Typography whith font size xsmall is passed 1`] = `
.c0 {
  font-weight: 400;
  color: #000000;
  font-size: 1.2rem;
}

<span
  class="c0"
  color="black"
>
  text
</span>
`;
